HCMS Objects refactoring

ku5ic opened the pull request:
## Changes
Implements the **Object Field** visual and interaction updates from the [Figma design](https://www.figma.com/design/4XsQ9YFfV1xVbLEvmJ3JDX/Webiny-UI?node-id=138-3708).  
Improves clarity and usability for repeatable object fields in complex forms.

<details>
  <summary>SCREENSHOTS</summary>
  
<img width="1194" height="1235" alt="Screenshot 2025-10-21 at 14 13 33" src="https://github.com/user-attachments/assets/13bc8b8e-b3bf-438f-bb7d-aa5226e7a95b" />


</details>

### User Story 1 — Visual distinction
- Added clear title, divider, and background per design.
- Conditional background logic:
  - **Level 1** (in grey form) → white background.
  - **Level 2** (nested) → grey background.
- Multi-value objects now have distinct color boxes for better separation.

### User Story 2 — Collapsible objects
- Added **collapse/expand** functionality for repeatable object sections.
- Improved visibility of collapsible areas (clearer than old UI).
- *Collapse All / Expand All* functionality is out of scope.

All existing functionality preserved.  
Closes [WEB-5371](https://webiny.atlassian.net/browse/WEB-5371)

---

## How Has This Been Tested?
- Manually tested nested and multi-value object scenarios.
- Verified background logic and collapse behavior.
- Compared against Figma for visual accuracy.
- Confirmed no regressions in existing form behavior.

---

## Documentation
No new documentation required.  
Changes align with existing **Object Field** behavior and may be mentioned in the UI changelog if part of a release.

ku5ic committed:
feat(cms): improve DynamicSection label styling

Enhanced the DynamicSection label UI by updating its layout and colors.
Label is now positioned absolutely, uses accent colors, and has improved
spacing for better visual distinction and clarity.

WEB-5371

ku5ic committed:
feat(cms): improve object field label styling

Updated the singleObjectInline renderer to enhance the visual appearance of
object field labels. Added accent border, spacing, and absolute positioning
for the label, improving clarity and consistency in the UI.

WEB-5371

ku5ic committed:
refactor(cms): update Accordion usage and remove old component

Replaced custom Accordion implementation with @webiny/admin-ui Accordion
in MultiValueItemContainer. Removed the obsolete Accordion.tsx file.
Updated icon imports and button variants for consistency.

WEB-5371

ku5ic committed:
feat(cms): expand multi-value item by default if isExpanded

Set Accordion's defaultOpen prop based on isExpanded, allowing multi-value
items to be expanded by default when specified. Improves UX for object
field renderers in admin CMS.

ku5ic committed:
refactor(admin-ui): remove background prop from Accordion

Removed the `background` prop from the Accordion component and its
usages across the codebase. Background color is now determined by the
accordion's nesting depth using a context provider.

WEB-5371

ku5ic committed:
Merge branch 'next' into ku5ic/feature/WEB-5371/hcms_objects_refactoring

ku5ic committed:
refactor(admin-ui): extract accordion background logic

- Create DEPTH_BACKGROUNDS constant for background class names
- Extract getBackgroundByDepth helper function
- Rename depth variable to currentDepth for clarity
- Improve code readability and maintainability

WEB-5371

ku5ic committed:
Merge branch 'next' into ku5ic/feature/WEB-5371/hcms_objects_refactoring

ku5ic committed:
refactor(admin-ui): update Accordion component styling

- Add rounded corners to container variant
- Add left border accent and primary indicator bar to items
- Reorganize icon rendering logic in AccordionTrigger
- Add left padding to AccordionContent
- Remove bottom border from AccordionItem
- Remove AccordionItemAction separator
- Consolidate icon rendering with renderIcon helper function

- Remove fieldsWrapperStyle from CMS object field renderer
These changes improve the visual hierarchy and consistency of
accordion items across the admin interface.

WEB-5371

ku5ic committed:
refactor(admin-ui): extract accordion trigger classes to variables

Extract inline Tailwind classes into named constants for better
readability and maintainability in AccordionTrigger component.

WEB-5371

ku5ic committed:
refactor(admin-ui): simplify AccordionContent component

- Refactor AccordionContent to arrow function with implicit return
- Consolidate CSS classes for better readability
- Add right padding (wby-pr-xxl) to content container
- Split animation classes into separate array items
- Destructure className prop explicitly
- Remove unnecessary comment about design tokens

WEB-5371

ku5ic committed:
refactor(admin-ui): extract accordion props splitting logic

Extract props splitting logic from useMemo into a separate function
and move constant classes array outside component to improve code
organization and readability.

- Extract `splitAccordionProps` helper function
- Move `ACCORDION_ITEM_CLASSES` to module scope
- Simplify `AccordionItemBase` component structure
- Maintain same functionality with cleaner code organization

WEB-5371

ku5ic committed:
refactor(admin-ui): simplify Accordion component code

- Remove unused DEPTH_BACKGROUNDS constant object
- Inline background color strings in getBackgroundByDepth function
- Simplify currentDepth calculation by removing intermediate variable
- Clean up trailing whitespace in variant definition
- Refactor exports to use named export for Accordion component

WEB-5371

ku5ic committed:
fix(app-headless-cms): add separator between accordion actions

Add a separator between the move up/down button and delete button
in the MultiValueItemContainer to improve visual distinction between
actions.

WEB-5371

ku5ic committed:
refactor(admin-ui): remove visual indicator from AccordionTrigger

Remove the primary-colored vertical bar element from the
AccordionTrigger component, simplifying its visual design.

WEB-5371

ku5ic committed:
feat(headless-cms): replace emoji with SVG icon for validation

Replace the emoji-based error icon with a custom SVG warning triangle
icon for better visual consistency and cross-platform compatibility.

The new icon:
- Uses a proper SVG with a red circular background
- Maintains the same spacing and positioning
- Removes the CSS variable customization option
- Provides consistent rendering across all browsers and OS

BREAKING CHANGE: The `--error-icon` CSS variable is no longer
supported for customizing the validation error icon.

WEB-5371

ku5ic committed:
refactor(cms): improve validation indicators code quality

- Extract constants to uppercase naming convention
- Encapsulate SVG data URL creation in a function
- Create reusable mixin helper function
- Extract validation marker logic into focused functions:
  * markFieldAsInvalid: Adds validation class to specific field
  * markParentFieldsAsInvalid: Recursively marks parent fields
  * clearValidationMarkers: Removes all validation classes
  * applyValidationMarkers: Applies validation to all invalid fields
- Consolidate duplicate CSS selectors
- Improve code comments for better readability
- Remove duplicate margin-right property in error icon styles
- Add className to useEffect dependency array for correctness

No functional changes, only code organization and maintainability
improvements.

WEB-5371

ku5ic committed:
fix(headless-cms): adjust validation error icon positioning

- Position error icons absolutely to prevent layout shifts
- Hide error icons on object/objects field group labels
- Set specific positioning (left: -10px, top: 17px) for accordion
  and title text elements to maintain consistent alignment

WEB-5371

ku5ic committed:
fix(headless-cms): add border styling to invalid object fields

Add red border styling to object fields with the "objects" renderer
when they are in an invalid state. This improves visual feedback by
making validation errors more noticeable for object field types.

WEB-5371

ku5ic committed:
Merge branch 'next' into ku5ic/feature/WEB-5371/hcms_objects_refactoring

ku5ic committed:
refactor(headless-cms): improve validation indicators code quality

- Rename ERROR_BACKGROUND to ERROR_COLOR_MUTED for clarity
- Add return type annotation to createErrorIconDataUrl
- Replace string-based mixins with emotion css helper
- Simplify validation marker utilities with better composition
- Extract getFieldElement and getParentPaths helper functions
- Improve code organization and readability
- Remove unnecessary createMixin wrapper function
- Clean up spacing and formatting inconsistencies

WEB-5371

Pavel910 commented on packages/admin-ui/src/Accordion/components/AccordionTrigger.tsx:
```diff
@@ -44,15 +68,21 @@ const AccordionTrigger = ({
+                    <div className={"wby-ml-sm wby-w-[20px] wby-h-[1px] wby-bg-neutral-strong"} />
+                )}
                 {draggable ? <AccordionItemDragHandle /> : null}
+                {renderIcon(isInteractable)}
```
This line is inconsistent with the line above. 
If you have `{draggable ? <AccordionItemDragHandle /> : null}` (a ternary operator), than why go for a utility function in the line below? Turn the utility function into a proper component, like `InteractiveIcon` or whatever, and then do `{isInteractable ? <InteractiveIcon/> : null}`. It's easier to read, requires no jumps in the code to understand, and follows the existing coding style.

Pavel910 commented on packages/admin-ui/src/Accordion/components/AccordionTrigger.tsx:
```diff
@@ -44,15 +68,21 @@ const AccordionTrigger = ({
+                    cursorClass
                 )}
             >
+                {!icon && !isInteractable && (
+                    <div className={"wby-ml-sm wby-w-[20px] wby-h-[1px] wby-bg-neutral-strong"} />
+                )}
```
Always use a proper ternary operator, and return `null`. The way you have it here will return `false`, and it can be visible in the UI. The proper React way of doing this is toreturn `null`, which means `render nothing`.

ku5ic replied:
Addressed with [b313ccf](https://github.com/webiny/webiny-js/pull/4747/commits/b313ccfbe44c78dcd8f64c3447f19cccfe3eeac6)

Pavel910 commented on packages/admin-ui/src/Accordion/components/AccordionTrigger.tsx:
```diff
@@ -71,20 +101,7 @@ const AccordionTrigger = ({
-                                icon={<KeyboardArrowDownIcon />}
-                            />
-                        )}
+                        {renderIcon(!isInteractable)}
```
Same as before, use a proper component.

ku5ic replied:
Addressed with [b313ccf](https://github.com/webiny/webiny-js/pull/4747/commits/b313ccfbe44c78dcd8f64c3447f19cccfe3eeac6)

Pavel910 commented on packages/app-headless-cms/src/admin/plugins/fieldRenderers/DynamicSection.tsx:
Have you verified that this didn't break anything in other renderers that uas `DynamicSection`? This component is used across various renderers. Review and confirm that this change didn't break other fields that use this component.

ku5ic replied:
Confirmed. 👍

Pavel910 reviewed (changes requested):
Left a few change requests.

ku5ic committed:
refactor(admin-ui): extract OpenCloseIndicator component

Extract inline icon rendering logic into a dedicated
OpenCloseIndicator component for better readability and reusability.

- Create OpenCloseIndicator component with show prop
- Remove renderIcon inline function
- Replace renderIcon calls with OpenCloseIndicator component
- Maintain same visual behavior and conditional rendering

WEB-5371

ku5ic committed:
refactor(admin-ui): simplify OpenCloseIndicator component

Remove conditional rendering logic from OpenCloseIndicator component
by eliminating the `show` prop. Move the visibility control to the
parent component using conditional rendering, making the code more
idiomatic and easier to understand.

WEB-5371

ku5ic committed:
Merge branch 'next' into ku5ic/feature/WEB-5371/hcms_objects_refactoring

adrians5j commented on packages/admin-ui/src/Accordion/components/AccordionTrigger.tsx:
```diff
@@ -71,20 +100,7 @@ const AccordionTrigger = ({
-                                icon={<KeyboardArrowDownIcon />}
-                            />
-                        )}
+                        {!isInteractable ? <OpenCloseIndicator /> : null}
```
I saw we lost the separator. Shouldn't this still be here?

Another question: I can see before we were rendering OpenCloseIndicator if `interactive===true`, now it's the opposite?

adrians5j commented on packages/admin-ui/src/Accordion/Accordion.tsx:
```diff
@@ -2,22 +2,18 @@ import React from "react";
-            underline: "wby-accordion-variant-underline "
-        },
-        background: {
-            base: "wby-accordion-background-base",
```
Why did we loose bg prop? We still need it per DS: https://www.figma.com/design/f0QUDWX37Kt5X53eltTRiT/Webiny-Design-System?node-id=3253-1490&t=KDI8H2G0VBoKW34F-4

adrians5j reviewed (commented):
Left a couple of comments code-related, but maybe the visual comments I'm about too post are results of the changes, so maybe we just focus on these first.

#### 1. `background`
I guess the main problem is the removal of background prop. Hence, all the accordions in storybook are gray, but, per DS, they need to be [white](https://www.figma.com/design/f0QUDWX37Kt5X53eltTRiT/Webiny-Design-System?node-id=3253-1490&t=KDI8H2G0VBoKW34F-4).

<img width="978" height="704" alt="Image" src="https://github.com/user-attachments/assets/de8b6a8c-6d49-492e-859b-ee5b52903171" />

#### 2. Weird line on the left
Another thing I don't understand is this line here:

<img width="912" height="466" alt="Image" src="https://github.com/user-attachments/assets/362baff4-9dd0-4b84-9aff-da47bbb1c8c4" />

I don't see this anywhere in DS.

#### 3. Chevron position
Why was this chevron moved to left? Again something I don't see in DS.

<img width="794" height="467" alt="Image" src="https://github.com/user-attachments/assets/f89fb193-4234-49b1-9e6f-d754b9b663c7" />

---

It's worth mentioning that Figma ([DS](https://www.figma.com/design/f0QUDWX37Kt5X53eltTRiT/Webiny-Design-System?node-id=3245-2246&t=KDI8H2G0VBoKW34F-0)) and our components and their code code (Storybook) must always be in sync (the more the better). 

So, if we're creating a screen and we see a variation of a component that's not in the DS / our code, then we need to go back: talk to Kreso about updating the component in DS / go back to the component code and update it. And while doing that, testing the change in storybook is a good way to do it. Basically, whenever I'm adding a feature into a component, I just start of by creating a new story and test my changes there. That way I can test the feature, plus we have the feature instantly documented.

You probably know all of these things already, but yeah, doesn't hurt to revisit.
